src/main/java/dev/sigstore/KeylessVerificationResult.java

```diff
@@ -0,0 +1,128 @@
+ */
+package dev.sigstore;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import dev.sigstore.json.GsonSupplier;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class KeylessVerificationResult {
+
+  enum Check {
+    NONE("none"),
+    SIGNATURE("Signature is valid for artifact and public key"),
+    FULCIO_CERT("Fulcio certificate chains up to fulcio root"),
+    FULCIO_CERT_SCT("Fulcio certificate sct is signed by the ct log"),
+    REKOR_ENTRY_EXISTS("Rekor entry was found on the log"),
+    REKOR_ENTRY_SIGNED("Rekor entry is signed by the rekor public key"),
+    REKOR_ENTRY_CONTENTS("Rekor entry body content matches request"),
+    REKOR_ENTRY_INCLUSION("Rekor entry inclusion proof verified"),
+    SIGNATURE_TIME("Signing occurred while certificate was valid");
+
+    private final String description;
+
+    Check(String description) {
+      this.description = description;
+    }
+
+    public String getDescription() {
+      return description;
+    }
+  }
+
+  private final List<Check> passingChecks;
+  private final Check failedCheck;
+  private final String failedReason;
+
+  public KeylessVerificationResult(
+      ImmutableList<Check> passingChecks, Check failedCheck, String failedReason) {
+    this.passingChecks = passingChecks;
+    this.failedCheck = failedCheck;
+    this.failedReason = failedReason;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder {
+    private final List<Check> passingChecks = new ArrayList<>();
+
+    private Builder() {}
+
+    public Builder checkPassed(Check check) {
+      if (passingChecks.contains(check)) {
+        throw new IllegalStateException(
+            "Passing check: " + check + " is already in passing checks list");
+      }
+      passingChecks.add(check);
+      return this;
+    }
+
+    public KeylessVerificationResult checkFailed(Check check, String reason) {
+      if (passingChecks.contains(check)) {
+        throw new IllegalStateException(
+            "Failing check: " + check + " is also in passing checks list");
+      }
+      return new KeylessVerificationResult(ImmutableList.copyOf(passingChecks), check, reason);
+    }
+
+    public KeylessVerificationResult allChecksPassed() {
+      return new KeylessVerificationResult(ImmutableList.copyOf(passingChecks), Check.NONE, null);
+    }
+  }
+
+  public boolean isVerified() {
+    return failedCheck == Check.NONE;
+  }
+
+  public Optional<Check> getFailedCheck() {
+    if (failedCheck == Check.NONE) {
+      return Optional.empty();
+    }
+    return Optional.ofNullable(failedCheck);
+  }
+
+  public Optional<String> getFailedReason() {
+    return Optional.ofNullable(failedReason);
+  }
+
+  public List<Check> getPassingChecks() {
+    return passingChecks;
+  }
+
+  public String generateJsonReport() {
```

Comment:
is this like the thing you see on the cosign CLI after you call verify successfully?

Reply:
Yeah maybe the tooling that uses this library could generate a report for someone if something failed. I dunno